In [1]:
from datetime import datetime
import pandas as pd

Так как API перестал работать, воспользовался ручным копированием информации об оставшихся играх Ови в 2023-2024 годах

In [2]:
df = pd.read_csv('Сезон 23-24.csv', sep=';')

In [3]:
df

date opponent  goals  assists  points  plusMinus  penaltyMinutes  \
0     Apr 5, 24    @ CAR      2        0       2          1               0   
1    Mar 24, 24   vs WPG      2        0       2          2               0   
2    Mar 20, 24   vs TOR      2        0       2          1               2   
3    Mar 18, 24    @ CGY      2        0       2          0               0   
4    Feb 20, 24   vs NJD      2        1       3          1               2   
..          ...      ...    ...      ...     ...        ...             ...   
105  Jan 17, 23   vs MIN      0        2       2          0               0   
106  Jan 16, 23    @ NYI      0        1       1          1               0   
107  Jan 14, 23   vs PHI      1        0       1          0               0   
108  Jan 11, 23    @ PHI      0        0       0         -2               2   
109   Jan 8, 23   vs CBJ      0        0       0          0               0   

     powerPlayGoals  powerPlayPoints  shortHandedGoals  gameWinningGoals  \
0                 0                0                 0                 0   
1                 0                0                 0                 0   
2                 1                1                 0                 0   
3                 2                2                 0                 1   
4                 1                1                 0                 0   
..              ...              ...               ...               ...   
105               0                0                 0                 0   
106               0                0                 0                 0   
107               0                0                 0                 0   
108               0                0                 0                 0   
109               0                0                 0                 0   

     overTimeGoals  shots  shifts timeOnIce  
0                0      4      22     19:37  
1                0      5      29     21:37  
2                0      2      24     15:06  
3                0      5      23     16:51  
4                0      5      23     17:59  
..             ...    ...     ...       ...  
105              0      3      22     21:41  
106              0      2      24     21:25  
107              0      5      20     21:50  
108              0      6      22     20:43  
109              0      5      21     19:46  

[110 rows x 15 columns]

Дальнейшие этапы:  
1. Приведение формата датасета к уже имеющеемуся, дата, команда соперника, добавление параметра домашней игры и др.
2. Склейка двух датасетов и проверка по количеству игр

In [4]:
def parse_date(x):
    return datetime.strptime(x, '%b %d, %y').strftime('%Y-%m-%d')

In [6]:
df['date']

0       Apr 5, 24
1      Mar 24, 24
2      Mar 20, 24
3      Mar 18, 24
4      Feb 20, 24
          ...    
105    Jan 17, 23
106    Jan 16, 23
107    Jan 14, 23
108    Jan 11, 23
109     Jan 8, 23
Name: date, Length: 110, dtype: object

In [7]:
df['date']=df['date'].apply(parse_date)

In [8]:
df

date opponent  goals  assists  points  plusMinus  penaltyMinutes  \
0    2024-04-05    @ CAR      2        0       2          1               0   
1    2024-03-24   vs WPG      2        0       2          2               0   
2    2024-03-20   vs TOR      2        0       2          1               2   
3    2024-03-18    @ CGY      2        0       2          0               0   
4    2024-02-20   vs NJD      2        1       3          1               2   
..          ...      ...    ...      ...     ...        ...             ...   
105  2023-01-17   vs MIN      0        2       2          0               0   
106  2023-01-16    @ NYI      0        1       1          1               0   
107  2023-01-14   vs PHI      1        0       1          0               0   
108  2023-01-11    @ PHI      0        0       0         -2               2   
109  2023-01-08   vs CBJ      0        0       0          0               0   

     powerPlayGoals  powerPlayPoints  shortHandedGoals  gameWinningGoals  \
0                 0                0                 0                 0   
1                 0                0                 0                 0   
2                 1                1                 0                 0   
3                 2                2                 0                 1   
4                 1                1                 0                 0   
..              ...              ...               ...               ...   
105               0                0                 0                 0   
106               0                0                 0                 0   
107               0                0                 0                 0   
108               0                0                 0                 0   
109               0                0                 0                 0   

     overTimeGoals  shots  shifts timeOnIce  
0                0      4      22     19:37  
1                0      5      29     21:37  
2                0      2      24     15:06  
3                0      5      23     16:51  
4                0      5      23     17:59  
..             ...    ...     ...       ...  
105              0      3      22     21:41  
106              0      2      24     21:25  
107              0      5      20     21:50  
108              0      6      22     20:43  
109              0      5      21     19:46  

[110 rows x 15 columns]

In [10]:
df['isHome']=df['opponent'].apply(lambda x: x[-5:-4]=='s')

In [11]:
df

date opponent  goals  assists  points  plusMinus  penaltyMinutes  \
0    2024-04-05    @ CAR      2        0       2          1               0   
1    2024-03-24   vs WPG      2        0       2          2               0   
2    2024-03-20   vs TOR      2        0       2          1               2   
3    2024-03-18    @ CGY      2        0       2          0               0   
4    2024-02-20   vs NJD      2        1       3          1               2   
..          ...      ...    ...      ...     ...        ...             ...   
105  2023-01-17   vs MIN      0        2       2          0               0   
106  2023-01-16    @ NYI      0        1       1          1               0   
107  2023-01-14   vs PHI      1        0       1          0               0   
108  2023-01-11    @ PHI      0        0       0         -2               2   
109  2023-01-08   vs CBJ      0        0       0          0               0   

     powerPlayGoals  powerPlayPoints  shortHandedGoals  gameWinningGoals  \
0                 0                0                 0                 0   
1                 0                0                 0                 0   
2                 1                1                 0                 0   
3                 2                2                 0                 1   
4                 1                1                 0                 0   
..              ...              ...               ...               ...   
105               0                0                 0                 0   
106               0                0                 0                 0   
107               0                0                 0                 0   
108               0                0                 0                 0   
109               0                0                 0                 0   

     overTimeGoals  shots  shifts timeOnIce  isHome  
0                0      4      22     19:37   False  
1                0      5      29     21:37    True  
2                0      2      24     15:06    True  
3                0      5      23     16:51   False  
4                0      5      23     17:59    True  
..             ...    ...     ...       ...     ...  
105              0      3      22     21:41    True  
106              0      2      24     21:25   False  
107              0      5      20     21:50    True  
108              0      6      22     20:43   False  
109              0      5      21     19:46    True  

[110 rows x 16 columns]

In [12]:
teams_dict = {"ANA":'Anaheim Ducks', 
"ARI":'Arizona Coyotes',
 "ATL":'Atlanta Thrashers',
 "BOS":'Boston Bruins',
 "BUF":'Buffalo Sabres',
 "CGY":'Calgary Flames',
 "CAR":'Carolina Hurricanes',
 "CHI":'Chicago Blackhawks',
 "COL":'Colorado Avalanche',
 "CBJ":'Columbus Blue Jackets',
 "EDM":'Edmonton Oilers',
 "NJD":'New Jersey Devils',
 "PHI":'Philadelphia Flyers',
 "FLA":'Florida Panthers',
 "NYR":'New York Rangers',
 "NYI":'New York Islanders',
 "MTL":'Montréal Canadiens',
 "OTT":'Ottawa Senators',
 "DET":'Detroit Red Wings',
 "NSH":'Nashville Predators',
 "LAK":'Los Angeles Kings',
 "DAL":'Dallas Stars',
 "MIN":'Minnesota Wild',
 "PHX":'Phoenix Coyotes',
 "PIT":'Pittsburgh Penguins',
 "SJS":'San Jose Sharks',
 "SEA":'Seattle Kraken',
 "STL":'St. Louis Blues',
 "TBL":'Tampa Bay Lightning',
 "TOR":'Toronto Maple Leafs',
 "VAN":'Vancouver Canucks',
 "VGK":'Vegas Golden Knights',
 "WPG":'Winnipeg Jets'}

In [15]:
df['opponent']=df['opponent'].apply(lambda x:teams_dict[x[-3:]])

In [16]:
df

date               opponent  goals  assists  points  plusMinus  \
0    2024-04-05    Carolina Hurricanes      2        0       2          1   
1    2024-03-24          Winnipeg Jets      2        0       2          2   
2    2024-03-20    Toronto Maple Leafs      2        0       2          1   
3    2024-03-18         Calgary Flames      2        0       2          0   
4    2024-02-20      New Jersey Devils      2        1       3          1   
..          ...                    ...    ...      ...     ...        ...   
105  2023-01-17         Minnesota Wild      0        2       2          0   
106  2023-01-16     New York Islanders      0        1       1          1   
107  2023-01-14    Philadelphia Flyers      1        0       1          0   
108  2023-01-11    Philadelphia Flyers      0        0       0         -2   
109  2023-01-08  Columbus Blue Jackets      0        0       0          0   

     penaltyMinutes  powerPlayGoals  powerPlayPoints  shortHandedGoals  \
0                 0               0                0                 0   
1                 0               0                0                 0   
2                 2               1                1                 0   
3                 0               2                2                 0   
4                 2               1                1                 0   
..              ...             ...              ...               ...   
105               0               0                0                 0   
106               0               0                0                 0   
107               0               0                0                 0   
108               2               0                0                 0   
109               0               0                0                 0   

     gameWinningGoals  overTimeGoals  shots  shifts timeOnIce  isHome  
0                   0              0      4      22     19:37   False  
1                   0              0      5      29     21:37    True  
2                   0              0      2      24     15:06    True  
3                   1              0      5      23     16:51   False  
4                   0              0      5      23     17:59    True  
..                ...            ...    ...     ...       ...     ...  
105                 0              0      3      22     21:41    True  
106                 0              0      2      24     21:25   False  
107                 0              0      5      20     21:50    True  
108                 0              0      6      22     20:43   False  
109                 0              0      5      21     19:46    True  

[110 rows x 16 columns]

In [17]:
df.to_pickle('./season23_24.pkl')

In [18]:
new_df = pd.read_pickle("./ovechkin.pkl")

In [19]:
new_df

date               opponent goals assists points plusMinus  \
2005020004 2005-10-05  Columbus Blue Jackets     2       0      2         1   
2005020020 2005-10-07      Atlanta Thrashers     0       1      1        -2   
2005020029 2005-10-08      Atlanta Thrashers     0       1      1         0   
2005020042 2005-10-10       New York Rangers     1       0      1         1   
2005020054 2005-10-12    Carolina Hurricanes     1       0      1         0   
...               ...                    ...   ...     ...    ...       ...   
2022020566 2022-12-29        Ottawa Senators     1       0      1         1   
2022020580 2022-12-31     Montréal Canadiens     3       0      3         2   
2022020599 2023-01-03         Buffalo Sabres     2       0      2         2   
2022020617 2023-01-05  Columbus Blue Jackets     1       0      1         0   
2022020622 2023-01-06    Nashville Predators     0       0      0         0   

           powerPlayGoals powerPlayPoints powerPlayTimeOnIce shortHandedGoals  \
2005020004              1               1              06:18                0   
2005020020              0               1              07:14                0   
2005020029              0               1              07:30                0   
2005020042              1               1              04:32                0   
2005020054              0               0              04:32                0   
...                   ...             ...                ...              ...   
2022020566              0               0              01:38                0   
2022020580              1               1              05:55                0   
2022020599              0               0              02:00                0   
2022020617              1               1              01:35                0   
2022020622              0               0              05:53                0   

           gameWinningGoals overTimeGoals shots penaltyMinutes shifts  \
2005020004                0             0     5              2     21   
2005020020                0             0     1              0     22   
2005020029                0             0     3              4     21   
2005020042                1             0     6              2     18   
2005020054                0             0     6              0     17   
...                     ...           ...   ...            ...    ...   
2022020566                0             0     6              0     24   
2022020580                0             0     5              2     23   
2022020599                0             0     7              0     25   
2022020617                0             0     2              0     20   
2022020622                0             0     4              0     19   

           timeOnIce  isHome  isWin   isOT    season  
2005020004     17:36    True   True  False  20052006  
2005020020     17:44    True  False  False  20052006  
2005020029     16:21   False  False  False  20052006  
2005020042     15:03    True   True  False  20052006  
2005020054     13:53   False  False  False  20052006  
...              ...     ...    ...    ...       ...  
2022020566     21:17    True  False   True  20222023  
2022020580     19:31    True   True  False  20222023  
2022020599     20:33    True  False   True  20222023  
2022020617     18:08   False   True  False  20222023  
2022020622     22:29    True  False  False  20222023  

[1316 rows x 20 columns]

In [20]:
new_df_1 = pd.read_pickle("./season23_24.pkl")
new_df_1

date               opponent  goals  assists  points  plusMinus  \
0    2024-04-05    Carolina Hurricanes      2        0       2          1   
1    2024-03-24          Winnipeg Jets      2        0       2          2   
2    2024-03-20    Toronto Maple Leafs      2        0       2          1   
3    2024-03-18         Calgary Flames      2        0       2          0   
4    2024-02-20      New Jersey Devils      2        1       3          1   
..          ...                    ...    ...      ...     ...        ...   
105  2023-01-17         Minnesota Wild      0        2       2          0   
106  2023-01-16     New York Islanders      0        1       1          1   
107  2023-01-14    Philadelphia Flyers      1        0       1          0   
108  2023-01-11    Philadelphia Flyers      0        0       0         -2   
109  2023-01-08  Columbus Blue Jackets      0        0       0          0   

     penaltyMinutes  powerPlayGoals  powerPlayPoints  shortHandedGoals  \
0                 0               0                0                 0   
1                 0               0                0                 0   
2                 2               1                1                 0   
3                 0               2                2                 0   
4                 2               1                1                 0   
..              ...             ...              ...               ...   
105               0               0                0                 0   
106               0               0                0                 0   
107               0               0                0                 0   
108               2               0                0                 0   
109               0               0                0                 0   

     gameWinningGoals  overTimeGoals  shots  shifts timeOnIce  isHome  
0                   0              0      4      22     19:37   False  
1                   0              0      5      29     21:37    True  
2                   0              0      2      24     15:06    True  
3                   1              0      5      23     16:51   False  
4                   0              0      5      23     17:59    True  
..                ...            ...    ...     ...       ...     ...  
105                 0              0      3      22     21:41    True  
106                 0              0      2      24     21:25   False  
107                 0              0      5      20     21:50    True  
108                 0              0      6      22     20:43   False  
109                 0              0      5      21     19:46    True  

[110 rows x 16 columns]

In [21]:
result = pd.concat([new_df,new_df_1])
result

date               opponent goals assists points  \
2005020004  2005-10-05 00:00:00  Columbus Blue Jackets     2       0      2   
2005020020  2005-10-07 00:00:00      Atlanta Thrashers     0       1      1   
2005020029  2005-10-08 00:00:00      Atlanta Thrashers     0       1      1   
2005020042  2005-10-10 00:00:00       New York Rangers     1       0      1   
2005020054  2005-10-12 00:00:00    Carolina Hurricanes     1       0      1   
...                         ...                    ...   ...     ...    ...   
105                  2023-01-17         Minnesota Wild     0       2      2   
106                  2023-01-16     New York Islanders     0       1      1   
107                  2023-01-14    Philadelphia Flyers     1       0      1   
108                  2023-01-11    Philadelphia Flyers     0       0      0   
109                  2023-01-08  Columbus Blue Jackets     0       0      0   

           plusMinus powerPlayGoals powerPlayPoints powerPlayTimeOnIce  \
2005020004         1              1               1              06:18   
2005020020        -2              0               1              07:14   
2005020029         0              0               1              07:30   
2005020042         1              1               1              04:32   
2005020054         0              0               0              04:32   
...              ...            ...             ...                ...   
105                0              0               0                NaN   
106                1              0               0                NaN   
107                0              0               0                NaN   
108               -2              0               0                NaN   
109                0              0               0                NaN   

           shortHandedGoals gameWinningGoals overTimeGoals shots  \
2005020004                0                0             0     5   
2005020020                0                0             0     1   
2005020029                0                0             0     3   
2005020042                0                1             0     6   
2005020054                0                0             0     6   
...                     ...              ...           ...   ...   
105                       0                0             0     3   
106                       0                0             0     2   
107                       0                0             0     5   
108                       0                0             0     6   
109                       0                0             0     5   

           penaltyMinutes shifts timeOnIce  isHome  isWin   isOT    season  
2005020004              2     21     17:36    True   True  False  20052006  
2005020020              0     22     17:44    True  False  False  20052006  
2005020029              4     21     16:21   False  False  False  20052006  
2005020042              2     18     15:03    True   True  False  20052006  
2005020054              0     17     13:53   False  False  False  20052006  
...                   ...    ...       ...     ...    ...    ...       ...  
105                     0     22     21:41    True    NaN    NaN       NaN  
106                     0     24     21:25   False    NaN    NaN       NaN  
107                     0     20     21:50    True    NaN    NaN       NaN  
108                     2     22     20:43   False    NaN    NaN       NaN  
109                     0     21     19:46    True    NaN    NaN       NaN  

[1426 rows x 20 columns]

1426 - Именно столько игр у Ови на данный момент, что совпадает с количеством строк в нашем датасете  
URL для проверки - https://www.nhl.com/capitals/player/alex-ovechkin-8471214